Add new rule `no-action-on-submit-button`

JoaoDsv opened the pull request:
Fixes https://github.com/ember-template-lint/ember-template-lint/issues/44.

It implements the rule proposal of `no-action-on-submit-button`.

# no-action-on-submit-button

This rule requires all `<button>` elements with a `type="submit"` attribute to not have any action.

When the `type` attribute of `<button>` elements is `submit`, the action should be on the `<form>` element instead of directly on the button.

By default, the `type` attribute of `<button>` elements is `submit`.

## Examples

This rule **forbids** the following:

```hbs
<form>
  <button type='submit' {{on 'click' this.handleClick}} />
  <button type='submit' {{action 'handleClick'}} />
  <button {{on 'click' this.handleClick}} />
  <button {{action 'handleClick'}} />
</form>
```

This rule **allows** the following:

```hbs
// In a <form>
<form>
  <button type='button' {{on 'click' this.handleClick}} />
  <button type='button' {{action 'handleClick'}} />
  <button type='submit' />
  <button />
</form>

// Outside a <form>
<button type='submit' {{on 'click' this.handleClick}} />
<button type='submit' {{action 'handleClick'}} />
<button {{on 'click' this.handleClick}} />
<button {{action 'handleClick'}} />
```

bertdeblock commented on test/unit/rules/no-action-on-submit-button-test.js:
```diff
@@ -0,0 +1,75 @@
+      result: {
+        message: ERROR_MESSAGE,
+        source: '<button {{action "handleClick"}} />',
+        moduleId: 'layout.hbs',
```
Don't think that `moduleId` is needed for these cases.

bmish replied:
Removed.

bertdeblock reviewed (commented):
Should we also check if the parent `form` already has an action defined and if so, assume that the action on the submit button is an arbitrary event? E.g. for tracking purposes. Or would this be to far-fetched?

bertdeblock commented on lib/rules/no-action-on-submit-button.js:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        let typeAttribute = attributes.find((it) => isTypeAttribe(it));
+        let onClickModifier = modifiers.find((it) => isOnClickModifier(it));
```
```js
let typeAttribute = attributes.find(isTypeAttribe);
```
Would do the same btw.

bmish replied:
The performance shouldn't be an issue, but yes this could be optimized if desired.

bmish commented on lib/rules/no-action-on-submit-button.js:
```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+const Rule = require('./base');
```
```suggestion
const Rule = require('./_base');
```

bmish commented on lib/rules/no-action-on-submit-button.js:
```diff
@@ -0,0 +1,69 @@
+
+module.exports = class NoActionOnSubmitButton extends Rule {
+  logNode({ node, message }) {
+    return this.log({
```
```suggestion
    return this.log({
      node,
```

bmish commented on test/unit/rules/no-action-on-submit-button-test.js:
```diff
@@ -0,0 +1,75 @@
+'use strict';
```
I added these tests.

bmish commented on lib/rules/no-action-on-submit-button.js:
```diff
@@ -0,0 +1,73 @@
+          return (
+            path.original === 'on' &&
+            params[0].value === 'click' &&
+            params[1].type === 'PathExpression'
```
Fixed.

bmish commented:
I fixed the conflicts, tests, and many (but not all) of the comments.

There's another comment here that may need to be addressed still: https://github.com/ember-template-lint/ember-template-lint/issues/44#issuecomment-218965427

bertdeblock commented:
If we need to check if the button has a `form` parent, we could make [this function](https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/rules/require-button-type.js#L10-L20) a shared helper and use that here as well? Though, if this rule specifically checks for `type="submit"` and buttons without a type (we also have the `require-button-type` rule to prevent this) that already seems like a good start?

bmish commented:
@JoaoDsv if you'd like to finish this PR up, let us know. Otherwise, I think @bertdeblock and I will try to finish it up. @bertdeblock, I'll give you access to push updates to this PR, so we can finish addressing comments including reducing false positives and improving consistency.

JoaoDsv commented:
Thanks for your feedbacks @bmish @bertdeblock !
Sorry for the late answer, I'm happy to try my best to address your comments

Feel free to commit at any time to help me to get this PR merged, it will be very welcome 🙂

JoaoDsv commented:
I addressed the remaining feedbacks ✅ 

Except reducing the false positives by checking the presence a parent `form` element. I'll handle it ASAP and let you know!

MelSumner commented on docs/rule/no-action-on-submit-button.md:
```diff
@@ -0,0 +1,35 @@
+<button type="button" {{on "click" this.handleClick}} />
+<button type="button" {{action "handleClick"}} />
+<button type="submit" />
+<button />
```
This would be disallowed by the `require-button-type` rule.

JoaoDsv replied:
Thanks for your comment @MelSumner. Sure, but aren't rules independent from each other?

My point is: should we take an other rule into account in the unit test of this current rule?

If users decide to disable the `require-button-type` rule, then this syntax will be allowed by the current rule, because it's not its role to prevent the missing type.

What is your opinion about it?

MelSumner replied:
So I try to write new rules with existing rules in mind; I don't want a rule I create to fail another existing rule. 

But @rwjblue might have some more clarity here. 

(PS overall I am excited about this rule!)

JoaoDsv replied:
Hi @MelSumner 

I'm back at work on this rule, which I'd forgotten for a long time 🙇 

This rule is not encouraging `<button />` without (explicit) type.  It's essentially a syntax that won't make this rule fail, so I listed it in the "valid" scenarios.

Would you prefer me to remove `<button />` from these scenarios?

- The downside is that this scenario won't be clearly mentioned while likely to happen.
- The advantage is that it won't encourage people using this syntax (not allowed by `require-button-type` dedicated rule)

bmish commented on test/unit/rules/no-action-on-submit-button-test.js:
```diff
@@ -0,0 +1,115 @@
+    {
+      template: '<button {{action this.handleClick}} />',
+
+      result: {
```
We have updated rules to use snapshot testing with `verifyResults()`. Mind switching to that?

JoaoDsv replied:
Sure, let me do it right now

bmish commented:
Looks good except we still need the parent `<form>` check.

bmish commented:
@JoaoDsv thanks for your continued work on this. Let us know when it's ready (with my last comment addressed).

JoaoDsv commented:
Hi @bmish @bertdeblock 

I have belatedly addressed your request about the parent `<form>` in this last commit: https://github.com/ember-template-lint/ember-template-lint/pull/1931/commits/24567e2f33a4b50a44c2f8178c0899593df9bb24. Is it what you had in mind?

ℹ️ Nicely spotted, as this also seems to solve @rwjblue's related concern: https://github.com/ember-template-lint/ember-template-lint/issues/44#issuecomment-218965427.

Would you mind having a last review and if everything looks ready, approve it?

bmish commented on test/unit/rules/no-action-on-submit-button-test.js:
```diff
@@ -0,0 +1,203 @@
+    '<button type="submit" {{on "click" (fn this.addNumber 123)}} />',
+  ],
+
+  bad: [
```
Shouldn't we add a test for a deeply-nested button like this to make sure we aren't only checking the direct parent?

* `<form><div><button {{action this.handleClick}} /></div></form>`

JoaoDsv replied:
Good idea! I added it as last scenario ✅

bmish commented:
@JoaoDsv just to confirm, we should enable this as `recommended` in the next major release, right?

JoaoDsv commented:
> @JoaoDsv just to confirm, we should enable this as `recommended` in the next major release, right?

I'd indeed suggest to:
- add the rule in the next minor release
- flag the rule as `recommended` in the next major release

By doing so, we additionally have some time to implement a `--fix` option to make the upgrade to the next major smoother. Wdyt @bmish?

bmish reviewed (approved):
Sounds good, thanks for your work to get this finished!
